Source/WTF/wtf/text/cocoa/TextStreamCocoa.mm

```diff
@@ -19,54 +19,79 @@
  */
 
 #import "config.h"
-#import <wtf/text/TextStream.h>
+#import <wtf/text/TextStreamCocoa.h>
 
+#import <objc/runtime.h>
 #import <wtf/text/cf/StringConcatenateCF.h>
 
 namespace WTF {
 
-TextStream& TextStream::operator<<(id object)
+TextStream& operator<<(TextStream& ts, id object)
 {
-    if ([object isKindOfClass:[NSArray class]])
-        return *this << static_cast<NSArray *>(object);
-
-    if ([object conformsToProtocol:@protocol(NSObject)])
-        m_text.append([object description]);
-    else
-        m_text.append("(id)");
-    return *this;
-}
+    if (object_isClass(object)) {
+        ts << NSStringFromClass(object);
+        return ts;
+    }
 
-TextStream& TextStream::operator<<(NSArray *array)
-{
-    *this << "[";
+    auto outputArray = [&](NSArray *array) {
+        ts << "[";
+
+        for (NSUInteger i = 0; i < array.count; ++i) {
+            id item = array[i];
+            ts << item;
+            if (i < array.count - 1)
+                ts << ", ";
+        }
+
+        ts << "]";
+    };
+
+    if ([object isKindOfClass:[NSArray class]]) {
+        outputArray(object);
+        return ts;
+    }
+
+    auto outputDictionary = [&](NSDictionary *dictionary) {
+        ts << "{";
+        bool needLeadingComma = false;
+
+        [dictionary enumerateKeysAndObjectsUsingBlock:[&](id key, id value, BOOL *) {
+            if (needLeadingComma)
+                ts << ", ";
+            needLeadingComma = true;
 
-    for (NSUInteger i = 0; i < array.count; ++i) {
-        id item = array[i];
-        *this << item;
-        if (i < array.count - 1)
-            *this << ", ";
+            ts << key;
+            ts << ": ";
+            ts << value;
+        }];
+
+        ts << "}";
+    };
+
+    if ([object isKindOfClass:[NSDictionary class]]) {
+        outputDictionary(object);
+        return ts;
```

Comment:
is this line needed, since the line underneath will return ts anyways?

Reply:
I think the explicit early return is good (someone might add another clause below this one).